extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/GenericImporter.java

```diff
@@ -0,0 +1,172 @@
+      }
+      return builder.toString();
+    }
+  }
+
+  ContainerSerializer<C, R> containerSerializer;
+  URL endpoint;
+  Monitor monitor;
+  AppCredentials appCredentials;
+  OkHttpClient client = new OkHttpClient();
+  ObjectMapper om = new ObjectMapper();
+  Map<UUID, OAuthTokenManager> jobTokenManagerMap = new HashMap<>();
+
+  static final MediaType JSON = MediaType.parse("application/json");
+
+  public GenericImporter(
+      ContainerSerializer<C, R> containerSerializer,
+      AppCredentials appCredentials,
+      URL endpoint,
+      Monitor monitor) {
+    this.monitor = monitor;
+    this.appCredentials = appCredentials;
+    this.endpoint = endpoint;
+    this.containerSerializer = containerSerializer;
+    configureObjectMapper(om);
+  }
+
+  @VisibleForTesting
+  static void configureObjectMapper(ObjectMapper objectMapper) {
+    // ZonedDateTime and friends
+    objectMapper.registerModule(new JavaTimeModule());
+    // Optional fields
+    objectMapper.registerModule(new Jdk8Module());
+    // ISO timestamps rather than unix epoch seconds
+    objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+  }
+
+  @Override
+  public ImportResult importItem(
+      UUID jobId,
+      IdempotentImportExecutor idempotentExecutor,
+      TokensAndUrlAuthData initialAuthData,
+      C data)
+      throws Exception {
+    OAuthTokenManager tokenManager =
+        jobTokenManagerMap.computeIfAbsent(
+            jobId,
+            ignored -> new OAuthTokenManager(initialAuthData, appCredentials, client, monitor));
+    for (ImportableData<R> importableData : containerSerializer.apply(data)) {
+      idempotentExecutor.executeAndSwallowIOExceptions(
+          importableData.getIdempotentId(),
+          importableData.getName(),
+          () ->
+              tokenManager.withAuthData(
+                  authData -> importSingleItem(jobId, authData, importableData)));
+    }
+    return new ImportResult(ResultType.OK);
+  }
+
+  boolean parseResponse(Response response) throws IOException, InvalidTokenException {
+    if (response.code() >= 400) {
+      byte[] body = response.body().bytes();
+      ErrorResponse error;
+      try {
+        error = om.readValue(body, ErrorResponse.class);
+      } catch (JsonParseException | JsonMappingException e) {
+        throw new IOException(
+            format(
+                "Unexpected response (%d) '%s'",
+                response.code(), new String(body, StandardCharsets.UTF_8)),
+            e);
+      }
+
+      if (response.code() == 401 && error.getError().equals("invalid_token")) {
+        throw new InvalidTokenException(error.toString(), null);
+      } else {
+        throw new IOException(format("Error (%d) %s", response.code(), error.toString()));
+      }
+    }
+    if (response.code() < 200 || response.code() >= 300) {
+      throw new IOException(format("Unexpected response code (%d)", response.code()));
+    }
+    return true;
```

Comment:
this implies that we are okay with 200 response code while in the documentation it seems like we are suggesting the API should return 201?

Reply:
We should be OK with a 200 response code.  That's a more normal response to a POST request anyway, and it indicates success without promising that a HTTP resource was created.  A 201 response code comes with expectations that the responder will include Location headers, which may not be appropriate depending on what the responder has done with the content sent.  From the perspective of the sender/exporter being DTP, either response code is considered a success.

Reply:
Casting a big net for success responses was intentional, to try to follow the robustness principle. Agreed that the docs don't need to be so prescriptive, so I've updated them 👍